src/Bicep.Core/Semantics/DeclarationVisitor.cs

```diff
@@ -161,33 +163,54 @@ public override void VisitProviderDeclarationSyntax(ProviderDeclarationSyntax sy
-            {
-                declaredType = ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).UnrecognizedProvider(syntax.Specification!.Name));
-            }
-            else
-            {
-                declaredType = namespaceType;
-            }
+                if (!features.ExtensibilityEnabled)
+                {
+                    return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).ProvidersAreDisabled());
+                }
 
-            var symbol = new ProviderNamespaceSymbol(this.context, syntax, declaredType);
-            DeclareSymbol(symbol);
+                // Check for interpolated specification strings
+                if (syntax.SpecificationString is StringSyntax specificationString && specificationString.IsInterpolated())
+                {
+                    return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax.SpecificationString).ProviderSpecificationInterpolationUnsupported());
+                }
+
+                if (!syntax.Specification.IsValid)
+                {
+                    return (syntax.SpecificationString is StringSyntax)
+                        ? ErrorType.Create(DiagnosticBuilder.ForPosition(syntax.Specification).InvalidProviderSpecification())
+                        : ErrorType.Empty();
+                }
+
+                // Check if the MSGraph provider is recognized and enabled
+                if (syntax.Specification.Name == MicrosoftGraphNamespaceType.BuiltInName && !features.MicrosoftGraphPreviewEnabled)
+                {
+                    return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).UnrecognizedProvider(syntax.Specification.Name));
+                }
+
+                Uri? providerUri = null;
+                if (syntax.Specification.Name == AzNamespaceType.BuiltInName &&
+                    features.DynamicTypeLoadingEnabled &&
+                    !this.artifactFileLookup.TryGetProviderFileUri(syntax).IsSuccess(out providerUri, out var providerUriLookupErrorBuilder))
+                {
+                    return ErrorType.Create(providerUriLookupErrorBuilder(DiagnosticBuilder.ForPosition(syntax)));
+                }
+
+                ResourceTypesProviderDescriptor providerDescriptor = new(
+                    syntax.Specification.Name,
+                    syntax.Alias?.IdentifierName,
+                    providerUri,
+                    syntax.Specification.Version);
+
+                if (namespaceProvider.TryGetNamespace(providerDescriptor, targetScope, features, sourceFileKind) is not { } namespaceType)
+                {
+                    return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).UnrecognizedProvider(syntax.Specification!.Name));
+                }
+
+                return namespaceType;
+            }
+            DeclareSymbol(new ProviderNamespaceSymbol(this.context, syntax, resolveProviderSymbol()));
```

Comment:
The benefit of having a helper function is to simplify the control flow, we can simply return the result without needing to evaluate the other conditionals since the conditionals are all mutually exclusive.

Reply:
Method-local functions have a runtime cost, since they will be reinstantiated on every method invocation and capture the local scope in their closure. Unless `resolveProviderSymbol` needs to close over method-local state, it would be more efficient to declare it as an instance function.